flink-streaming-java/src/test/java/org/apache/flink/streaming/api/DataStreamTest.java

```diff
@@ -906,6 +919,256 @@ public void testChannelSelectors() {
+
+		input.keyBy(keySelector);
+	}
+
+	// composite key tests : POJOs
+
+	@Test
+	public void testPOJONestedArrayKeyRejection() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<POJOwithHashCode> input = env.fromElements(
+				new POJOwithHashCode(new int[] {1, 2}));
+
+		TypeInformation<?> expectedTypeInfo = new TupleTypeInfo<Tuple1<int[]>>(
+				PrimitiveArrayTypeInfo.INT_PRIMITIVE_ARRAY_TYPE_INFO);
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+		expectedException.expectMessage("This type (" + expectedTypeInfo + ") cannot be used as key.");
+
+		input.keyBy("id");
+	}
+
+	@Test
+	public void testNestedArrayWorkArround() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<POJOwithHashCode> input = env.fromElements(
+				new POJOwithHashCode(new int[] {1, 2}));
+
+		input.keyBy(new KeySelector<POJOwithHashCode, POJOwithHashCode>() {
+			@Override
+			public POJOwithHashCode getKey(POJOwithHashCode value) throws Exception {
+				return value;
+			}
+		}).addSink(new SinkFunction<POJOwithHashCode>() {
+			@Override
+			public void invoke(POJOwithHashCode value) throws Exception {
+				Assert.assertEquals(value.getId(), new int[]{1, 2});
+			}
+		});
+	}
+
+	@Test
+	public void testPOJOnoHashCodeKeyRejection() {
+
+		KeySelector<POJOnoHashCode, POJOnoHashCode> keySelector =
+				new KeySelector<POJOnoHashCode, POJOnoHashCode>() {
+					@Override
+					public POJOnoHashCode getKey(POJOnoHashCode value) throws Exception {
+						return value;
+					}
+				};
+
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<POJOnoHashCode> input = env.fromElements(
+				new POJOnoHashCode(new int[] {1, 2}));
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+
+		input.keyBy(keySelector);
+	}
+
+	// composite key tests : Tuples
+
+	@Test
+	public void testTupleNestedArrayKeyRejection() {
+		StreamExecutionEnvironment env = StreamExecutionEnvironment.getExecutionEnvironment();
+
+		DataStream<Tuple2<Integer[], String>> input = env.fromElements(
+				new Tuple2<>(new Integer[] {1, 2}, "test-test"));
+
+		TypeInformation<?> expectedTypeInfo = new TupleTypeInfo<Tuple2<Integer[], String>>(
+				BasicArrayTypeInfo.INT_ARRAY_TYPE_INFO, BasicTypeInfo.STRING_TYPE_INFO);
+
+		// adjust the rule
+		expectedException.expect(InvalidProgramException.class);
+		expectedException.expectMessage("This type (" + expectedTypeInfo + ") cannot be used as key.");
+
+		input.keyBy(new KeySelector<Tuple2<Integer[],String>, Tuple2<Integer[],String>>() {
+			@Override
+			public Tuple2<Integer[], String> getKey(Tuple2<Integer[], String> value) throws Exception {
+				return value;
+			}
+		});
+	}
+
+	@Test
+	public void testPrimitiveKeyRejection() throws Exception {
```

Comment:
The name is totally misleading, you are right. This test only verifies that we did not break anything and that even if you leave the default signature for the `keySelector` (returning `Object` although your key is of a different type) it still works.